ga4gh/datamodel/sequenceAnnotations.py

```diff
@@ -0,0 +1,432 @@
+        :param pageSize: int representing number of records to return
+        :param referenceName: string representing reference name, ex 'chr1'
+        :param start: int position on reference to start search
+        :param end: int position on reference to end search >= start
+        :param parentId: string restrict search by id of parent node.
+        :return an array of dictionaries, representing the returned data.
+        """
+        # TODO: Refactor out common bits of this and the above count query.
+        sql = (
+            "SELECT * FROM FEATURE WHERE "
+            "reference_name = ? "
+            "AND end > ? "  # compare this to query start
+            "AND start < ? ")  # and this to query end
+        # TODO: Optimize by refactoring out string concatenation
+        sql_args = (referenceName, start, end)
+        if parentId is not None:
+            sql += "AND parent_id = ? "
+            sql_args += (parentId,)
+        if featureTypes is not None and len(featureTypes) > 0:
+            sql += "AND ontology_term IN ("
+            sql += ", ".join(["?", ] * len(featureTypes))
+            sql += ") "
+            sql_args += tuple(featureTypes)
+        sql += sqliteBackend.limitsSql(pageToken, pageSize)
+        query = self._dbconn.execute(sql, sql_args)
+        return sqliteBackend.sqliteRows2dicts(query.fetchall())
+
+    def getFeatureById(self, featureId):
+        """
+        Fetch feature by featureID.
+
+        :param featureId: the FeatureID as found in GFF3 records
+        :return: dictionary representing a feature object,
+            or None if no match is found.
+        """
+        sql = ("SELECT * FROM FEATURE WHERE id = ?")
+        query = self._dbconn.execute(sql, (featureId,))
+        ret = query.fetchone()
+        if ret is None:
+            return None
+        return sqliteBackend.sqliteRow2Dict(ret)
+
+
+# Problem: FeatureIds from GFF3 often contain colons,
+# and the CompoundId mechanism reserves colons for its special separator
+# character. Thus the following pair of functions that mangles and
+# de-mangles featureId's to allow them to work in compoundIds.
+def _decolonize(featureId):
+    """
+    Replaces colons in input with CompoundId-safe character
+    :param featureId: string potentially with colons in it
+    :return: same string with colons replaced with safe character
+    """
+    return featureId.replace(':', ';')
+
+
+def _recolonize(mangled_featureId):
+    """
+    Replaces CompoundId-safe character in input with colons
+    :param featureId: string potentially with replacement character
+    :return: same string with colons in place.
+    """
+    return mangled_featureId.replace(';', ':')
+
+
+class AbstractFeatureSet(datamodel.DatamodelObject):
+    """
+    A set of sequence features annotations
+    """
+    compoundIdClass = datamodel.FeatureSetCompoundId
+
+    def __init__(
+            self, parentContainer, localId, referenceSetId=None):
```

Comment:
It appears there is not currently a way to associate a featureSet with a reference set. This is not unique to sequence annotations, however, and can be left as an issue.